### PR TITLE
Add check for image name length to not exceed `255`

### LINKED ISSUE #2542: ImageName should check image name length while validating image name
### Component

JKube Kit

### Description
ImageName seems to parse any image name no matter how long it is. This is not aligned with Docker behavior.

Docker seems to have a limit of 255 characters in image repository's name (see [reference.go](https://github.com/distribution/reference/blob/8507c7fcf0da9f570540c958ea7b972c30eeaeca/reference.go#L39)). it throws this exception when image name length is above 255 characters:
```golang

// NameTotalLengthMax is the maximum total number of characters in a repository name.
NameTotalLengthMax = 255
ErrNameTooLong = fmt.Errorf("repository name must not be more than %v characters", NameTotalLengthMax)
```

It seems to be checked here in [reference.go](https://github.com/distribution/reference/blob/8507c7fcf0da9f570540c958ea7b972c30eeaeca/reference.go#L193).

We should add a check in ImageName's `doValidate` method to verify `repository`'s length is less than 255
### Expected Behavior
ImageName throws exception when image repository name length  is greater than 255

### Acceptance Criteria
- [ ] Validation check is added for image repository length in ImageName

### PR DESCRIPTION
**Pull request by Devashishbasu**
fixes #2542

### CODE REVIEW

**Comment by manusa** [comment]
### Eclipse JKube [CI Report](https://github.com/jkubeio/ci)

Started new GH workflow run for https://github.com/eclipse/jkube/pull/2555 (_2024-01-15T07:34:31Z_)

:gear: [JKube E2E Tests (7525694825)](https://github.com/jkubeio/ci/actions/runs/7525694825) :hourglass_flowing_sand:
- :hourglass_flowing_sand: [Bootstrap CI test run (#2555)](https://github.com/jkubeio/ci/actions/runs/7525694825/job/20482458006)

**Comment by rohanKanojia** [comment]
@Devashishbasu : Could you please uncomment this test?

https://github.com/eclipse/jkube/blob/c387c2be37561f4a07788de0ce3aa63ed2a8b4fc/jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameDistributionReferenceTest.java#L97-L98

**Comment by Devashishbasu** [comment]
> @Devashishbasu : Could you please uncomment this test?
> 
> https://github.com/eclipse/jkube/blob/c387c2be37561f4a07788de0ce3aa63ed2a8b4fc/jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameDistributionReferenceTest.java#L97-L98

do i need to also uncomment the line 98 ?

**Comment by rohanKanojia** [comment]
@Devashishbasu : Yes, but in the PR where you fix that issue.

**Comment by Devashishbasu** [comment]
> @Devashishbasu : Yes, but in the PR where you fix that issue.

done

**Comment by Devashishbasu** [comment]
> @Devashishbasu : Yes, but in the PR where you fix that issue.

when i uncommented the build got failed ...how can i fix this??

**Comment by Devashishbasu** [comment]
@rohanKanojia @manusa still stuck with this test case "repo@sha256:ffffffffffffffffffffffffffffffffff" can you please review  where i am making mistake?

**Comment by rohanKanojia** [comment]
@Devashishbasu : Are you sure you should be uncommenting that test case in this PR? 

In https://github.com/eclipse/jkube/pull/2555#issuecomment-1892163287 I meant uncommenting it in https://github.com/eclipse/jkube/pull/2557 not in this pull request.

**Comment by Devashishbasu** [comment]
> @Devashishbasu : Are you sure you should be uncommenting that test case in this PR?
> 
> In [#2555 (comment)](https://github.com/eclipse/jkube/pull/2555#issuecomment-1892163287) I meant uncommenting it in #2557 not in this pull request.

thankyou so much for clearing my doubt

File: jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/ImageName.java [hunk 1]
```diff
@@ -288,6 +294,11 @@ private void doValidate() {
                         checks[i], value, checkPattern.pattern()));
             }
         }
+
+        if (repository.length() > REPO_NAME_MAX_LENGTH) {
+            errors.add(String.format("Repository name must not be more than %d characters", REPO_NAME_MAX_LENGTH));
```
Comment (rohanKanojia) [inline jkube-kit/config/image/src/main/java/org/eclipse/jkube/kit/config/image/ImageName.java:299]: Could you please add one more test in ImageNameTest for verifying exception with this exact message is thrown when too long image name is provided?

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java [hunk 2]
```diff
@@ -123,6 +123,22 @@ private void verifyData(Object[] data) {
...
+        assertThatIllegalArgumentException()
+            .as("Too long image name should fail")
+            .isThrownBy(() -> new ImageName(tooLongName))
+            .withMessageContaining("Repository name must not be more than 255 characters");
+    }
+
+    private String generateTooLongImageName() {
+        StringBuilder tooLongName = new StringBuilder();
+        int maxLength = 255 + 1; // exceeding the maximum length
+        for (int i = 0; i < maxLength; i++) {
```
Comment (rohanKanojia) [inline jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java:138]: You can replace this by a prebuilt method from Apache's [StringUtils repeat](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#repeat-java.lang.String-int-)

File: jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java [hunk 3]
```diff
@@ -123,6 +123,22 @@ private void verifyData(Object[] data) {
     @Test
     void testIllegalFormat() {
         assertThrows(IllegalArgumentException.class, () -> new ImageName(""));
+
+        // New test for too long repository name
+        String tooLongName = generateTooLongImageName();
+        assertThatIllegalArgumentException()
```
Comment (rohanKanojia) [inline jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java:129]: Sorry, I meant to add a new test; not modify existing test. Could you please move this code block into a separate test?

```java
@Test 
void shouldThrowExceptionOnTooLongImageName() {

}
```

Comment (Devashishbasu) [reply jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java:131]: > Sorry, I meant to add a new test; not modify existing test. Could you please move this code block into a separate test?
> 
> ```java
> @Test 
> void shouldThrowExceptionOnTooLongImageName() {
> 
> }
> ```

Hii @rohanKanojia did all the changes you told me do, but getting a build failure

[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:38 min
[INFO] Finished at: 2024-01-22T14:44:48Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.1.2:test (default-test) on project jkube-kit-remote-dev: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/jenkins/agent/workspace/Build_Java8_PR-2555/jkube-kit/remote-dev/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.1.2:test (default-test) on project jkube-kit-remote-dev: There are test failures.

is it due to my changes? or can you suggest how can i fix this?

Comment (Devashishbasu) [reply jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java:131]: > > Sorry, I meant to add a new test; not modify existing test. Could you please move this code block into a separate test?
> > ```java
> > @Test 
> > void shouldThrowExceptionOnTooLongImageName() {
> > 
> > }
> > ```
> 
> Hii @rohanKanojia did all the changes you told me do, but getting a build failure
> 
> [INFO] BUILD FAILURE [INFO] ------------------------------------------------------------------------ [INFO] Total time: 03:38 min [INFO] Finished at: 2024-01-22T14:44:48Z [INFO] ------------------------------------------------------------------------ [ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.1.2:test (default-test) on project jkube-kit-remote-dev: There are test failures. [ERROR] [ERROR] Please refer to /home/jenkins/agent/workspace/Build_Java8_PR-2555/jkube-kit/remote-dev/target/surefire-reports for the individual test results. [ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream. [ERROR] -> [Help 1] org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.1.2:test (default-test) on project jkube-kit-remote-dev: There are test failures.
> 
> is it due to my changes? or can you suggest how can i fix this?

test added by me is passing

Comment (rohanKanojia) [reply jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java:131]: @Devashishbasu : This failure seems unrelated to your changes. Probably it's a flaky test that needs to be fixed. Could you please try pushing again and see if you see the issue again?

Comment (Devashishbasu) [reply jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java:131]: > @Devashishbasu : This failure seems unrelated to your changes. Probably it's a flaky test that needs to be fixed. Could you please try pushing again and see if you see the issue again?

Hi @rohanKanojia i've pushed the code again and seeing the same issue again...it is the same issue which you have opened  #2576

Comment (Devashishbasu) [reply jkube-kit/config/image/src/test/java/org/eclipse/jkube/kit/config/image/ImageNameTest.java:131]: > @Devashishbasu : This failure seems unrelated to your changes. Probably it's a flaky test that needs to be fixed. Could you please try pushing again and see if you see the issue again?

Hii @rohanKanojia ...still the same issue persists

**Comment by Devashishbasu** [comment]
Hii @manusa can you please review it?

**Comment by rohanKanojia** [comment]
@Devashishbasu : Your branch does not contain the commit where I disabled the failing test on  jdk8.  You need to rebase this branch with the latest updates in master.

**Comment by rohanKanojia** [comment]
@Devashishbasu : Please squash your commits into one commit.

**Pull request closed without merging**